### PR TITLE
MinIO file operations

### PR DESCRIPTION
**Pull request by GiancarloFusiello**
In reference to https://github.com/libero/shared-services-experiment/issues/3:

 - create environment for running a script of review operations

### CODE REVIEW

File: minio/scripts/file-operations.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+# Executable that waits for a port at an address to be open.
+# For further details go to: https://github.com/ufoscout/docker-compose-wait
+/wait
+
+echo "Add s3 location alias called storage to minio config"
+mc config host add storage http://minio:9000 longkey verysecretkey
+
+echo "Create private bucket (private by default)"
+mc mb storage/private-bucket
```
Comment (giorgiosironi) [inline minio/scripts/file-operations.sh]: weren't the operation supposed to use the `aws` cli? There would be some that are out of scope, but it demonstrates S3 compatibility

Comment (giorgiosironi) [reply minio/scripts/file-operations.sh]: (unless `mc` could be pointed to a real S3 bucket for comparison)

File: minio/scripts/file-operations.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+echo "Upload file to private bucket"
+mc cp ./assets/file.txt storage/private-bucket
+
+echo "Get meta data about uploaded file"
+mc stat storage/private-bucket/file.txt
```
Comment (giorgiosironi) [inline minio/scripts/file-operations.sh]: misses the similar: store arbitrary object tags onto a file

File: minio/scripts/file-operations.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+
+set -e
+
+# Executable that waits for a port at an address to be open.
+# For further details go to: https://github.com/ufoscout/docker-compose-wait
+/wait
+
+echo "Add s3 location alias called storage to minio config"
+mc config host add storage http://minio:9000 longkey verysecretkey
```
Comment (diversemix) [inline minio/scripts/file-operations.sh]: Should this use the env vars `${MINIO_ACCESS_KEY}`

File: minio/scripts/file-operations.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+echo "Create private bucket (private by default)"
```
Comment (giorgiosironi) [inline minio/scripts/file-operations.sh:13]: misses the similar "Create public bucket (files are read-only, no listings)"

File: minio/scripts/file-operations.sh (hunk 1 above)
Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: It's not mentioned in the ticket but shouldn't be an issue as the commands are very similar. I'll swap `mc` for `awscli`.

Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: updated. Please review.

File: minio/scripts/file-operations.sh (hunk 2 above)
Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: should be there now.

File: minio/scripts/file-operations.sh (hunk 4 above)
Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh:13]: should be there now.

File: minio/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,6 @@
+FROM mikesir87/aws-cli:latest
```
Comment (giorgiosironi) [inline minio/Dockerfile]: maybe `Dockerfile.aws-cli` to make clear this is not the server's Dockerfile, but the client's one?

File: minio/Makefile [hunk 6]
```diff
@@ -0,0 +1,8 @@
+stop:
+	docker-compose down -v
+
+file-ops:
+	docker-compose up file-ops
```
Comment (giorgiosironi) [inline minio/Makefile]: the `command` doesn't stay up, but exits, so this should be a `docker-compose run` command

File: minio/Makefile [hunk 7]
```diff
@@ -0,0 +1,8 @@
+stop:
+	docker-compose down -v
```
Comment (giorgiosironi) [inline minio/Makefile:2]: there are no actual volumes attached, so `-v` is unnecessary (and in general you wouldn't expect `stop` to remove them right?)

File: minio/policy.json [hunk 8]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (giorgiosironi) [inline minio/policy.json]: rename the file to `public-policy.json` or similar to communicate what it does?

File: minio/Makefile (hunk 7 above)
Comment (GiancarloFusiello) [reply minio/Makefile:2]: Leaving it in there doesn't provide any extra overhead

File: minio/policy.json (hunk 8 above)
Comment (GiancarloFusiello) [reply minio/policy.json]: done

File: minio/Makefile (hunk 6 above)
Comment (GiancarloFusiello) [reply minio/Makefile]: done.

File: minio/Dockerfile (hunk 5 above)
Comment (GiancarloFusiello) [reply minio/Dockerfile]: done.

File: minio/scripts/file-operations.sh [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+# Executable that waits for a port at an address to be open.
+# For further details go to: https://github.com/ufoscout/docker-compose-wait
+/wait
+
+echo "Create private bucket (private by default)"
+aws_cli s3 mb s3://private-bucket
+
+echo "Upload file to private bucket with custom meta data"
+aws_cli s3 cp ./assets/file.txt s3://private-bucket --content-type text/plain --metadata CustomMetaKey=CustomMetaValue
```
Comment (giorgiosironi) [inline minio/scripts/file-operations.sh]: it's good to be able to set metadata, but the key names have a [special meaning](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-metadata) that can lead a file to be encrypted or to produce a redirect to something else. What we want for arbitrary key-value pairs is `aws s3api put-object-tagging`, or an equivalent in `aws s3 cp` if it has an option like that (couldn't find it).

File: minio/scripts/file-operations.sh [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+aws_cli s3 cp ./assets/file.txt s3://public-bucket --content-type text/plain --metadata CustomMetaKey=CustomMetaValue
+
+echo "Get meta data about uploaded file"
+aws_cli s3api head-object --bucket public-bucket --key file.txt
+
+echo "Get file from public bucket without credentials"
+wget http://minio:9000/public-bucket/file.txt -O ./file1.txt
+
+echo "Clean up downloaded files"
+rm file.txt file1.txt
```
Comment (giorgiosironi) [inline minio/scripts/file-operations.sh]: more communicative names for these two files? `private-file.txt` vs `public-file.txt`?

File: minio/scripts/file-operations.sh (hunk 9 above)
Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: Looks like minio might not support this api yet: `An error occurred (NotImplemented) when calling the PutObjectTagging operation: A header you provided implies functionality that is not implemented`

Comment (giorgiosironi) [reply minio/scripts/file-operations.sh]: So we have to keep metadata as a fallback and avoid using keys that are reserved for AWS e.g. `x-amz` ones or `Content-Length`.

Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: adding to the conversation for record https://github.com/minio/minio/blob/master/cmd/generic-handlers.go#L473

Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: updating to use `x-amz-meta-mytagname`

Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: updated.

File: minio/scripts/file-operations.sh (hunk 10 above)
Comment (GiancarloFusiello) [reply minio/scripts/file-operations.sh]: updated.

**Review by giorgiosironi — APPROVED** [review]

**Pull request merged**